cpp/ql/src/semmle/code/cpp/ir/implementation/aliased_ssa/internal/SSAConstruction.qll

```diff
@@ -422,7 +422,11 @@ private predicate hasChiNode(Alias::VirtualVariable vvar, OldInstruction def) {
     defLocation.getVirtualVariable() = vvar and
     // If the definition totally (or exactly) overlaps the virtual variable, then there's no need for a `Chi`
     // instruction.
-    Alias::getOverlap(defLocation, vvar) instanceof MayPartiallyOverlap
+    (
+      Alias::getOverlap(defLocation, vvar) instanceof MayPartiallyOverlap or
+      def.getResultMemoryAccess() instanceof IndirectMayMemoryAccess or
```

Comment:
Why aren't these two new cases covered by the original `instanceof MayPartiallyOverlap` check?

Reply:
`getOverlap` computes overlaps between two `MemoryLocation`s, not between two accesses. This accounts for cases where the `MemoryLocation`s have an exact or total overlap, but the definition is not a must-write. (Note that we can't refactor `getOverlap` to take a`MemoryAccessKind` because some uses of `getOverlap` aren't in the context of a specific definition).

Reply:
OK, I may have to rethink how `getOverlap()` works, or at least how it's used. I think what you've done here is the right change for now. My intent was that a `MemoryLocation` could be a "may" access, in which case it would `MayPartiallyOverlap` anything that the equivalent "must" `MemoryLocation` overlaps in any way. If these last couple sentences give you any insight into a better fix, please go for it, but I'm fine with the changes you have already.